pilot/pkg/model/config_test.go

```diff
@@ -299,39 +300,101 @@ func TestResolveShortnameToFQDN(t *testing.T) {
 	}
 }
 
+func BenchmarkMostSpecificHostMatchExact(b *testing.B) {
+	length := 5000
+	m := make(map[host.Name]struct{})
+	var l []host.Name
+
+	for i := 0; i <= length; i++ {
+		h := host.Name(strconv.Itoa(i) + ".foo.bar.com")
+		l = append(l, h)
+		m[h] = struct{}{}
+	}
+
+	needle := host.Name(strconv.Itoa(length) + ".foo.bar.com")
+
+	b.ResetTimer()
+	for n := 0; n < b.N; n++ {
+		_, _ = model.MostSpecificHostMatch(needle, m, l)
```

Comment:
Don't you need to construct the map inside the loop? Of course a map has faster lookup than a list - but not if you need to allocate and populate a map each call like we are doing.